mlir/lib/Gradient/Transforms/PostprocessingPatterns.cpp

```diff
@@ -200,13 +205,74 @@ struct PostprocessReverseOp : public OpRewritePattern<ReverseOp> {
     }
 };
 
+struct RestoreReverseOp : public OpRewritePattern<ReverseOp> {
+    using OpRewritePattern<ReverseOp>::OpRewritePattern;
+
+    mlir::LogicalResult matchAndRewrite(ReverseOp op,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        // ReverseOp's output is optimized away by one-shot bufferize.
```

Comment:
Can you explain what you mean by this comment?

Reply:
Here is the example.
```mlir
gradient.reverse @bwd.rev(%arg0: memref<f64>) -> memref<f64>, memref<f64> attributes {argc = 2 : i64, implementation = @bwd, llvm.linkage = #llvm.linkage<internal>, resc = 1 : i64, tape = 0 : i64} {
    %0 = func.call @bwd(%arg0) : (memref<f64>) -> memref<f64>
    %alloc = memref.alloc() {alignment = 64 : i64} : memref<f64>
    memref.copy %0, %alloc : memref<f64> to memref<f64>
    gradient.return {empty = true} %alloc : memref<f64>
}
``` 
After `one-shot-bufferize`, we get:
```mlir
gradient.reverse @bwd.rev(%arg0: memref<f64>) -> memref<f64> attributes {argc = 2 : i64, implementation = @bwd, llvm.linkage = #llvm.linkage<internal>, resc = 1 : i64, tape = 0 : i64} {
    %0 = func.call @bwd(%arg0) : (memref<f64>) -> memref<f64>
    %alloc = memref.alloc() {alignment = 64 : i64} : memref<f64>
    memref.copy %0, %alloc : memref<f64> to memref<f64>
    gradient.return {empty = true} %alloc : memref<f64>
}
``` 
However, since `argc = 2`, we have to keep the return for Ezyme.